logging: uart: support multiple instances

ycsin opened the pull request:
Extends the UART log backend to support multiple UART instances.

Building `samples/hello_world` with `qemu_x86` using additional Kconfigs:

```kconfig
CONFIG_LOG=y
CONFIG_LOG_BACKEND_UART=y
```

## ROM reports:

### `main` branch

```
Path                              Size
=======================================
log_backend_uart.c (ROM)           208
├── char_out                        49
├── dropped                         16
├── format_set                      15
├── log_backend_uart                16
├── log_backend_uart_api            28
├── log_backend_uart_init            1
├── log_const_log_uart               8
├── log_output_uart                 16
├── panic                           23
└── process                         36
=======================================
log_backend_uart.c (RAM)            26
├── backend_cb_log_backend_uart      8
├── in_panic                         1
├── log_format_current               4
├── log_output_uart_control_block   12
└── uart_output_buf                  1
```

### This PR, using `zephyr,console`

```
Path                              Size
=======================================
log_backend_uart.c (ROM)           280
├── backend_cb_log_backend_uart0     8
├── char_out                        51
├── dropped                         22
├── format_set                      24
├── lbu_cb_ctx_0                    12
├── lbu_output_0                    16
├── log_backend_uart0               16
├── log_backend_uart_api            28
├── log_backend_uart_init           21
├── log_const_log_uart               8
├── panic                           29
└── process                         45
=======================================
log_backend_uart.c (RAM)            53
├── backend_cb_log_backend_uart0     8
├── lbu_buffer_0                     1
├── lbu_data_0                      32
└── lbu_output_0_control_block      12
```

|                         | main   |  PR    |      |
|-------------------------|--------|--------|------|
| RAM                     | 122912 | 122912 |      |
| ROM                     | 56856  | 56864  | +8   |
| `char_out`              | 49     | 51     | +2   |
| `dropped`               | 16     | 22     | +6   |
| `format_set`            | 15     | 24     | +9   |
| `log_backend_uart`      | 16     | 16     |      |
| `log_backend_uart_api`  | 28     | 28     |      |
| `log_backend_uart_init` |  1     | 21     | +20  |
| `log_const_log_uart`    |  8     |  8     |      |
| `log_output_uart`       | 16     | 16     |      |
| `panic`                 | 23     | 29     | +6   |
| `process`               | 36     | 45     | +9   |

### This PR, using `zephyr,log-uart`, 2 instances

```
Path                              Size
=======================================
log_backend_uart.c                 332
├── backend_cb_log_backend_uart0     8
├── backend_cb_log_backend_uart1     8
├── char_out                        51
├── dropped                         22
├── format_set                      24
├── lbu_cb_ctx_0                    12
├── lbu_cb_ctx_1                    12
├── lbu_output_0                    16
├── lbu_output_1                    16
├── log_backend_uart0               16
├── log_backend_uart1               16
├── log_backend_uart_api            28
├── log_backend_uart_init           21
├── log_const_log_uart               8
├── panic                           29
└── process                         45
=======================================
log_backend_uart.c (RAM)           106
├── backend_cb_log_backend_uart0     8
├── backend_cb_log_backend_uart1     8
├── lbu_buffer_0                     1
├── lbu_buffer_1                     1
├── lbu_data_0                      32
├── lbu_data_1                      32
├── lbu_output_0_control_block      12
└── lbu_output_1_control_block      12
```

cfriedt commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -103,27 +118,36 @@ static int char_out(uint8_t *data, size_t length, void *ctx)
-
 static void process(const struct log_backend *const backend,
-		union log_msg_generic *msg)
+		    union log_msg_generic *msg)
```
Should probably try to remove any pure formatting / space changes.

cfriedt commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -103,27 +118,36 @@ static int char_out(uint8_t *data, size_t length, void *ctx)
 
-static int format_set(const struct log_backend *const backend, uint32_t log_type)
+static int format_set(const struct log_backend *const backend,
+		      uint32_t log_type)
```
Same re formatting changes

cfriedt commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -1,49 +1,59 @@
 #include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_backend_std.h>
 #include <zephyr/logging/log_core.h>
 #include <zephyr/logging/log_output.h>
 #include <zephyr/logging/log_output_dict.h>
-#include <zephyr/logging/log_backend_std.h>
-#include <zephyr/logging/log.h>
-#include <zephyr/device.h>
-#include <zephyr/drivers/uart.h>
-#include <zephyr/sys/util_macro.h>
-#include <zephyr/sys/__assert.h>
 #include <zephyr/pm/device.h>
 #include <zephyr/pm/device_runtime.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util_macro.h>
```
This can probably be undone - better to avoid pure formatting changes

cfriedt commented on samples/hello_world/prj.conf:
```diff
@@ -1 +1,4 @@
 # nothing here
+CONFIG_LOG=y
```
You're already on it I'm sure, but please create separate samples and / or tests.

cfriedt commented on samples/hello_world/src/main.c:
```diff
@@ -5,9 +5,19 @@
  */
 
 #include <stdio.h>
+#include <zephyr/device.h>
```
I'm sure you're already on it, but please create separate samples and / or tests

cfriedt reviewed (commented):
Would be also good to add the chosen to docs 

https://docs.zephyrproject.org/latest/build/dts/api/api.html#zephyr-specific-chosen-nodes

cfriedt commented on dts/bindings/misc/zephyr,log-uarts.yaml:
```diff
@@ -0,0 +1,15 @@
+  uarts:
+    type: phandles
+    required: true
+    description: UART devices to be used by the UART log backend.
```
I believe descriptions need to be in "piped" format now "|<newline>..."

Commit:
logging: uart: support multiple instances

Extends the log_backend_uart to support logging to multiple
UART instances.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

ycsin committed:
test: logging: add testcase for multi-instance log backend uart

Add testcase for the multi-instance log backend uart using
emulated uart.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -103,27 +118,36 @@ static int char_out(uint8_t *data, size_t length, void *ctx)
 static void process(const struct log_backend *const backend,
 		union log_msg_generic *msg)
 {
+	const struct lbu_cb_ctx *ctx = backend->cb->ctx;
```
at some point the `backend->cb->ctx` becomes NULL, no sure why it happens yet

ycsin replied:
Found it:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/132b821d-fb71-4c90-adf2-bfbda51120d0)

ycsin committed:
tests: logging: log_syst: set backend ctx to its cb ctx when enable

Set the backend's context to its control block's context
instead of NULL when doing `log_backend_enable`, as the log
backend UART uses that later.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -197,5 +227,36 @@ const struct log_backend_api log_backend_uart_api = {
+LBU_DEFINE(DT_CHOSEN(zephyr_console), 0);
+#endif
+
+/*
+ * #if DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_log_uart), zephyr_log_uart)
+ * DT_FOREACH_PROP_ELEM_SEP(DT_CHOSEN(zephyr_log_uart), uarts, LBU_PHA_FN, ());
+ * #else
+ * LBU_DEFINE(DT_CHOSEN(zephyr_log_uart), 0);
+ * #endif
+ */
```
#65001

cfriedt commented on tests/subsys/logging/log_backend_uart/multi.overlay:
```diff
@@ -0,0 +1,32 @@
+
+/ {
+	chosen {
+		zephyr,log-uart = &log_uarts;
```
I think my preference would be either to use phandles property in the chosen node, or to simply use a child DT node under each UART selected to be a log backend.

However, the phandles approach does not seem to work for in the chosen node.

Personally, I think the chosen + node + phandles is kind of overkill.

cfriedt replied:
Alternatively, if it were possible to support a phandles array in chosen, that would be kind of ideal, but it might require some additional devicetree hacks.

nordicjm reviewed (commented):
Can you do a build without these changes and with these changes (ARM cortex ideally) with 1 device and note the flash/RAM usage on both?

nordic-krch commented:
+1 to @nordicjm  suggestion. I guess there will be increase so it would be good to wrap instance getting in some macro so that there is no code addition when 1 uart is used (99.99% of cases).

ycsin commented:
@nordic-krch @nordicjm

Building `samples/hello_world` with `qemu_x86` using additional Kconfigs:

```kconfig
CONFIG_LOG=y
CONFIG_LOG_BACKEND_UART=y
```

## ROM reports:

### `main` branch

```
Path                              Size
=======================================
log_backend_uart.c                 208
├── char_out                        49
├── dropped                         16
├── format_set                      15
├── log_backend_uart                16
├── log_backend_uart_api            28
├── log_backend_uart_init            1
├── log_const_log_uart               8
├── log_output_uart                 16
├── panic                           23
└── process                         36
```

### This PR, using `zephyr,console`

```
Path                              Size
=======================================
log_backend_uart.c                 280
├── backend_cb_log_backend_uart0     8
├── char_out                        51
├── dropped                         22
├── format_set                      24
├── lbu_cb_ctx_0                    12
├── lbu_output_0                    16
├── log_backend_uart0               16
├── log_backend_uart_api            28
├── log_backend_uart_init           21
├── log_const_log_uart               8
├── panic                           29
└── process                         45
```

### This PR, using `zephyr,log-uart`, 2 instances

```
Path                              Size
=======================================
log_backend_uart.c                 332
├── backend_cb_log_backend_uart0     8
├── backend_cb_log_backend_uart1     8
├── char_out                        51
├── dropped                         22
├── format_set                      24
├── lbu_cb_ctx_0                    12
├── lbu_cb_ctx_1                    12
├── lbu_output_0                    16
├── lbu_output_1                    16
├── log_backend_uart0               16
├── log_backend_uart1               16
├── log_backend_uart_api            28
├── log_backend_uart_init           21
├── log_const_log_uart               8
├── panic                           29
└── process                         45
```

nordic-krch reviewed (approved):
Checked that for cortex-m and riscv32 overhead is even smaller. Looks ok for me.
